Loot table API v2, iteration 2

Juuxel opened the pull request:
**Edit (13 May 2022):** Now merged with 1.18.2 branch and official TAW module. Also includes an internal API for resource tracking ported from #1983.

**Edit (31 Jan 2022):** This PR has now been updated to 1.18.1, uses interface injection and also creates the TAW module. The multipurpose `LootTableLoadingCallback` was replaced with two events with clear purposes (replacing or modifying).

**Edit (27 Nov 2021):** After 18 months of working this API, I made another refactor and started using transitive AWs instead of static accessor classes. They'll be merged into the TAW module once that comes along, so this is a draft for now.

## What does the loot table API do?
The main use case of this API is adding items/loot pools to loot tables (example: modded seeds to vanilla grass) or modifying them in some other way. There are also extended and more convenient versions of the vanilla loot table/pool builders, like in v1.

This PR is also a replacement for #629 with a different approach. This one is also backwards compatible (I had some trouble with that using the same design).

## Changes from V1

This API uses modern Yarn names and avoids "magic interfaces" implemented on vanilla classes. It also drops some unnecessary functionality that is fully public in modern vanilla.

### Transitive access wideners instead of magic interfaces and static access classes

The magic interfaces `FabricLootSupplier` and `FabricLootPool` are replaced with **transitive access wideners**. I also dropped some functionality that is now available in vanilla (`LootJsonParser` and `LootEntryTypeRegistry`).

- The returned lists from V1 magic interfaces are now immutable. The mutability was not documented at all, and this is safer.
- The access wideners also expose the bonus rolls of a loot pool.
- The implementation is much simpler now for V1 as well:
  - `LootEntryTypeRegistry` no longer uses mixins but calls `Registry.register` directly.
  - `LootJsonParser` no longer uses reflection.

### Injected builder interfaces

In the new V2 module, there are `FabricLootPoolBuilder` and `FabricLootTableBuilder` to replace the V1 builders.

- They are now interfaces that are injected to `LootPool.Builder` and `LootTable.Builder`, respectively.
- Minor renames: `of(LootTable/Pool)` -> `copyOf`
- The loot pool builder now has methods for working with collections as parameters.
- All `copyFrom` methods are gone. They had weird and arbitrarily chosen boolean flag parameters, and their few use cases can be replicated using v2 API methods.

### Changes to events

`LootTableLoadingCallback` has been replaced with `LootTableEvents.REPLACE` and `LootTableEvents.MODIFY` which clearly separate the two different purposes. `REPLACE` runs before `MODIFY`.

- The callback loop is now stopped when a loot table is replaced. This isn't really a breaking change to v1 as that behaviour was not defined at all.
- You can access a loot table's source in the events. This can be used to only apply modifications to built-in resources and not user-overridden ones, for example.
  - `VANILLA` and `MOD` for built-in content
  - `DATA_PACK` for external data packs
  - `REPLACED` for loot tables created in `LootTableEvents.REPLACE` (naturally only exists for loot tables processed in `LootTableEvents.MODIFY`)
  - This is implemented via a new internal API in resource loader for resource source tracking. The `FabricResource` interface, which might be promoted to public API in the future, exposes the `ResourcePackSource` of the `ResourcePack` instance that created the `Resource`.

Juuxel committed:
Improve loot table API

Alternative to #629.

- Deprecates all classes and methods that use outdated Yarn names.
- Adds FabricLootTable and FabricLootTableBuilder to replace
  the LootSupplier naming variants.
- Deprecates LootEntryTypeRegistry and LootJsonParser
  as their functionality is exposed in vanilla now.
- Adds methods to FabricLootPoolBuilder for working
  with collections as builder parameters.
- FabricLootPool and FabricLootTable/Supplier now return immutable lists
  instead of modifiable fixed-size ones.

Still WIP: LootTableLoadingCallback uses the deprecated FabricLootSupplier.

Update fabric-loot-tables-v1/src/main/java/net/fabricmc/fabric/api/loot/v1/FabricLootTableBuilder.java

Co-authored-by: i509VCB <i509vcb@gmail.com>

Fix compilation

Create loot table API v2

Move incorrect classes and revert unnecessary change

Add test for replacing loot tables

Document FabricLootPools

builder() -> create()

Add accessor for LootPool.bonusRolls

Add loot pool builder method for bonus rolls

Use Blocks.DIRT.getLootTableId() instead of raw string in v1 loot test mod

Make links in deprecation docs cleaner

Make FabricLootPoolBuilder.copyFrom(pool, true) also copy bonus rolls

...and mention it in the javadoc

Remove copyFrom from v2 builders

It seems like a maintainability mess if Mojang ever decides
to extend loot tables, and the chosen boolean flags are arbitrary.
It also doesn't really have use cases apart from the internal use
in the copyOf methods, and even then users can replicate its functionality
with the other API methods.

Rename 'supplier' to 'table' in LootManagerMixin

Add 'stable' lifecycle to loot-table-api-v2, deprecate v1

Add internal comment for implementors about updates

Cancel all remaining callbacks when a loot table is replaced

Remove unused shadowed logger from LootManagerMixin

Migrate subproject versioning to new system

Start the AW migration

Update test mod

i509VCB commented on fabric-loot-tables-v1/src/main/java/net/fabricmc/fabric/api/loot/v1/event/LootTableLoadingCallback.java:
```diff
@@ -45,5 +45,6 @@
 			}
 	);
 
+	// TODO: What should I do about FabricLootSupplierBuilder here?
```
New event sadly :(

Juuxel replied:
Yeah, I don't think there's another way. I'll make the new event be in the v1 root package and register this event's invoker as a callback for it.

i509VCB commented on fabric-loot-tables-v1/src/main/java/net/fabricmc/fabric/api/loot/v1/FabricLootTable.java:
```diff
@@ -0,0 +1,55 @@
+ * Accessor methods for loot table properties.
+ * This interface is automatically implemented on all {@link LootTable} instances.
+ */
+public interface FabricLootTable {
```
Ideally this interface should need to be exposed in API, a helper class with static `getBlah` methods would suffice here and be more resilient to api changes.

Juuxel replied:
I think you mean this interface *shouldn't* be exposed in the API? For that, I can make a `LootTables` class (prefixed with Fabric if needed) with a bunch of getters, and maybe do the same to loot pools too.

i509VCB commented on fabric-loot-tables-v1/src/main/java/net/fabricmc/fabric/api/loot/v1/FabricLootTableBuilder.java:
```diff
@@ -0,0 +1,152 @@
+	protected FabricLootTableBuilder() {
+	}
+
+	private FabricLootTableBuilder(LootTable supplier) {
```
```suggestion
	private FabricLootTableBuilder(LootTable lootTable) {
```

liach commented on fabric-loot-table-api-v2/src/main/java/net/fabricmc/fabric/api/loot/v2/FabricLootPoolBuilder.java:
```diff
@@ -0,0 +1,181 @@
+	 *
+	 * @return the created builder
+	 */
+	public static FabricLootPoolBuilder builder() {
```
```java
FabricLootPoolBuilder.builder()
```
looks a bit weird in usage.

i509VCB replied:
I think in this case since the builder is being made it would become `FabricLootPoolBuilder.create()`

Juuxel replied:
I'll change these to `create`.

liach commented on fabric-loot-table-api-v2/src/main/java/net/fabricmc/fabric/api/loot/v2/FabricLootTableBuilder.java:
```diff
@@ -0,0 +1,151 @@
+	 *
+	 * @return the created builder
+	 */
+	public static FabricLootTableBuilder builder() {
```
Same, `FabricLootTableBuilder.builder()` sounds a bit weird

i509VCB commented on fabric-loot-table-api-v2/src/main/java/net/fabricmc/fabric/api/loot/v2/FabricLootPools.java:
```diff
@@ -0,0 +1,58 @@
+		return ImmutableList.copyOf(((LootPoolAccessor) pool).getEntries());
+	}
+
+	public static List<LootCondition> getConditions(LootPool pool) {
```
javadoc needs to explain the resulting lists are immutable

Juuxel replied:
Oops, forgot to document these ones. I'll add that explanation as well.

i509VCB commented on fabric-loot-table-api-v2/src/main/java/net/fabricmc/fabric/api/loot/v2/LootTableLoadingCallback.java:
```diff
@@ -0,0 +1,65 @@
+ */
+@FunctionalInterface
+public interface LootTableLoadingCallback {
+	Event<LootTableLoadingCallback> EVENT = EventFactory.createArrayBacked(
```
is there a possibility for future loot table events?

Juuxel replied:
There's nothing that comes to my mind at least, but any future ones can be PR'd later. (I think a top-level callback containing its own `EVENT` field is fine for now)

i509VCB commented on fabric-loot-table-api-v2/src/main/java/net/fabricmc/fabric/api/loot/v2/LootTableLoadingCallback.java:
```diff
@@ -0,0 +1,65 @@
+	 * @param lootManager     the loot manager
+	 * @param id              the loot table ID
+	 * @param tableBuilder    a builder of the loot table being loaded
+	 * @param setter          a loot table setter for completely replacing the loaded loot table
```
does setting the replacement cancel future notification or does it silently change return result. I imagine this could be an area of contention a mod or two may fight over.

Juuxel replied:
It silently changes the result, which is problematic IMO (and also according to @asiekierka who added it in #160). We could figure out a better solution now.

Juuxel replied:
I updated this to cancel future callbacks now. It's different behaviour than in v1's original implementation, but it's also completely undefined in v1. It doesn't make sense to continue the callback loop anyway.

i509VCB commented on fabric-loot-table-api-v2/src/main/java/net/fabricmc/fabric/api/loot/v2/FabricLootPoolBuilder.java:
```diff
@@ -0,0 +1,171 @@
+ * An extended version of {@link LootPool.Builder}.
+ */
+public class FabricLootPoolBuilder extends LootPool.Builder {
+	private final LootPoolBuilderAccessor access = (LootPoolBuilderAccessor) this;
```
I'd probably add an internal comment here about implementors needing to add any new methods to this and the copyOf method.

i509VCB commented on fabric-loot-table-api-v2/src/main/resources/fabric.mod.json:
```diff
@@ -0,0 +1,26 @@
+  "description": "Hooks for manipulating loot tables.",
+  "mixins": [
+    "fabric-loot-table-api-v2.mixins.json"
+  ]
```
This module now needs to define a module lifecycle:
For the new module you would add the following to the fabric.mod.json
```json
  "custom": {
    "fabric-api:module-lifecycle": "stable"
  }
```

liach commented:
Hmm, just curious, how are priorities assigned?

<details>
<summary>About why I don't use discordapp anymore</summary>
Discord is problematic. I don't know if my account is even there because it may get nuked when one of the guilds I'm in gets nuked. It causes extraneous distraught and has very few meaningful discussion compared to like github discussion, irc (no bouncer still so not using as of now), or mailing lists.
</details>

i509VCB commented:
I'll detail priorities in a bit, currently it is somewhat along the lines of urgency

maityyy commented:
Is there some way to replace pool entries? It seems that the current API does not have this flexibility and it came to me to use AccessWidener to get access to many private fields...

maityyy commented:
`ItemEntry` has `item` field and it is private. It is rather inconvenient to replace one pool entry with another, because you have to add your own accessors/duck interfaces/access wideners... Perhaps this PR wants to solve this problem.

maityyy commented:
Why is `LootTableLoadingCallback` not in the event subpackage? If we want to add another one, it will be a mess. Other modules always add events to their package, so I think it is bad to move away from this concept.

Juuxel commented:
@maityyy I don't really think it would be a mess. 1) There are only four other classes in that package. 2) Other new modules are also moving away from event subpackages. Some of them use a `*Events` class instead, but I decided to leave the structure the same from v1.

liach commented:
Should this target 1.17?

Juuxel commented:
I'm not sure what the general policy here is. It should be really easy to port, though, if needed.

Technici4n commented:
> Should this target 1.17?

I don't think it should target 1.17, we would probably want to backport it anyway.

PS: Yes I will review this soon:tm:.

modmuss50 commented:
Makes little diffrence now, the main reason why we suggested targgeting the stable release is it saves PRs having to be updated each week. It takes the same amount of effrot to forward port or backport.

liach commented:
> Makes little diffrence now, the main reason why we suggested targgeting the stable release is it saves PRs having to be updated each week. It takes the same amount of effrot to forward port or backport.

oh, notice that porting may be affected by modern java features

Technici4n reviewed (approved):
Looks fantastic.

maityyy commented:
Why is this a draft?

Juuxel commented:
I need to rework it again to have interface injection according to @Technici4n - and interface injection is months away from stabilising at the very least. *(Edit: this aged like milk)*

It also needs the transitive access widener module to replace `FabricLootTables` and `FabricLootPools`, which I don't really want to create in a PR that will be stuck in the "waiting-forever-for-toolchain-features" limbo. The merge conflicts would be too much if another PR created the module.

modmuss50 commented:
> and interface injection is months away from stabilising at the very least.

It should be mostly ready now. Just need to update loom to 0.11.

Juuxel committed:
Use interface injection

Juuxel committed:
Fix some issues

Juuxel committed:
Merge remote-tracking branch 'upstream/1.18' into loot-table-v1-improvements

# Conflicts:
#	gradle.properties

Juuxel committed:
Remove outdated bonusRolls test from LootTest

It's a vanilla feature now.

Juuxel committed:
Create transitive access widener module

Technici4n commented on fabric-loot-table-api-v2/src/main/java/net/fabricmc/fabric/api/loot/v2/FabricLootPoolBuilder.java:
```diff
@@ -0,0 +1,114 @@
+	 * @return this builder
+	 */
+	default LootPool.Builder with(LootPoolEntry entry) {
+		throw new UnsupportedOperationException("Unimplemented");
```
Could possibly use a more descriptive error message to avoid scaring users unfamiliar with this pattern? `"Implemented via mixin"` or so seems nicer.

Technici4n reviewed (approved):
Looks very good.

Juuxel committed:
Replace LootTableLoadingCallback with two events

Juuxel committed:
Use friendlier exception message

Juuxel committed:
Add loot table sources

Juuxel committed:
Add resource source tracking

Technici4n commented:
Now that the TAW module exists, this might be ready for a final (sigh) update?

Juuxel committed:
Add resource pack source for DefaultResourcePack$1 (anon resource)

Juuxel committed:
Add license header

Juuxel committed:
Make FabricResource an internal API in resource loader

Juuxel committed:
Remove my TAW module

Juuxel committed:
Merge branch '1.18.2' into loot-table-v1-improvements

# Conflicts:
#	fabric-resource-loader-v0/src/main/resources/fabric-resource-loader-v0.mixins.json
#	gradle.properties

Juuxel committed:
Add loot table-related TAWs

Juuxel committed:
Run CI

Juuxel committed:
Fix LootUtil.determineSource giving null values

Juuxel committed:
Clarify LootUtil comment

Player3324 reviewed (approved):
LGTM, thanks! (excl module vs pkg name as discussed)

Juuxel committed:
Rename loot-table-api => loot-api + minor comment changes

Juuxel committed:
Add README

Juuxel committed:
Fix mixin file name

Juuxel committed:
Use ImmutableMap.Builder instead of HashMap.computeIfAbsent in loot event impl

Juuxel committed:
Prefix accessor methods with fabric_ to prevent conflicts with loot v1

Juuxel committed:
Document mixins

Juuxel committed:
Document mixins more extensively

Juuxel committed:
Improve NRMMixin comments

Juuxel committed:
Change weird wording in FabricResourceImpl

Juuxel committed:
Merge remote-tracking branch 'upstream/1.18.2' into loot-table-v1-improvements

# Conflicts:
#	fabric-resource-loader-v0/src/main/resources/fabric-resource-loader-v0.mixins.json

Technici4n commented on fabric-loot-api-v2/src/main/java/net/fabricmc/fabric/api/loot/v2/LootTableEvents.java:
```diff
@@ -0,0 +1,85 @@
+/**
+ * Events for manipulating loot tables.
+ */
+public final class LootTableEvents {
```
Would be nice to clarify that replaced loot tables can be modified in the `MODIFY` event, maybe.

Technici4n commented on fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/mixin/resource/loader/ResourceMixin.java:
```diff
@@ -14,20 +14,16 @@
-	}
+import net.fabricmc.fabric.impl.resource.loader.FabricResource;
+
+// Even though the interface is injected, we must extend it here,
```
It's not injected anymore, is it?

Juuxel committed:
Minor updates

- Support new built-in mod respack source
- Fix ResourceMixin comment
- Add more docs to LootTableEvents.MODIFY
- Add package-info

Juuxel committed:
Add license header

modmuss50 reviewed (approved):
Looks awesome 👍 Thanks a lot for this. I will put it into last call.

Juuxel committed:
Merge branch '1.18.2' into loot-table-v1-improvements
